substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java

```diff
@@ -369,4 +370,48 @@ public MethodEntry ensureMethodEntryForDebugRangeInfo(DebugRangeInfo debugRangeI
     public List<MethodEntry> getMethods() {
         return methods;
     }
+
+    public int lowpc() {
+        assert isPrimary();
+        return primaryEntries.get(0).getPrimary().getLo();
+    }
+
+    public int lowpcDeopt() {
+        assert isPrimary();
+        assert includesDeoptTarget();
+        int lowpc = -1;
+        for (PrimaryEntry primaryEntry : primaryEntries) {
+            Range primary = primaryEntry.getPrimary();
+            if (primary.isDeoptTarget()) {
+                lowpc = primary.getLo();
+                break;
+            }
+        }
+        assert lowpc >= 0;
+        return lowpc;
+    }
+
+    public int hipc() {
+        assert isPrimary();
+        if (!includesDeoptTarget()) {
+            return primaryEntries.get(primaryEntries.size() - 1).getPrimary().getHi();
+        } else {
+            Range lastPrimary = null;
```

Comment:
Afraid not. The name PrimaryEntry is rather misleading. What that type actually represents is a top level compiled method. The list we are iterating is a list of all top level compiled methods in a given class. Of course the list may be empty. If not then the list is sorted by method code range within the code section. Any two methods lie in disjoint ranges. The non-deopt methods (potentially plural) form an initial segment of the list and all their ranges are contiguous. Deopt methods (potentially plural) only exist when there are non-deopt equivalents. Deopt methods form a trailing segment of the list and all their ranges are contiguous. If both leading and trailing segment are non-empty then there will be a discontinuity between the ranges for these two segments. That's because the totality of non-deopt methods for all classes are generated before the totality of deopt methods for all classes.

Reply:
It might be more convenient to map the two segments to two ArrayLists instead, then these methods would become trivial.
In places where you need them to be seen as one you just use `java.util.stream.Stream#concat`

Reply:
As I explained to Foivos I was wanting to avoid having an extra field and list in each class entry for the normal case where there are no deopt compiled methods. However, I guess I can create the deopt primaries list on demand and use a `null` list when there are no deopt methods. That also allows me to use the value of the list field as a way to get rid of boolean field `includesDeoptTarget`.